### ISSUE TITLE
run interface poor error handling

### ISSUE THREAD

**Issue by laroque** [project8/dragonfly#27]
If you try to do a run with a name that is too long for the database insert, the database service returns a useful message but the run_interface object does not pass that error info back to the user, making the situation difficult to debug. Should figure out where the error is being handled and be sure to pass the return code all the way back to the user.

**Comment by les67** [comment]
#30 #46

**Comment by les67** [comment]
A couple of notes: 

1. The maximum number of characters (not bytes) for run_name in the runs.run_table is 80; run_name is of type varchar(80).
2. Upon insertion of a run name with 81 characters:
```
2016-11-17T21:45:34[INFO    ] dripline.core.service(316) -> received a message
2016-11-17T21:45:34[DEBUG   ] dripline.core.service(363) -> Acknowledging message 1
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-17T21:45:34Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "payload": {"values": [], "run_name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"}}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-17T21:45:34Z', u'payload': {u'values': [], u'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'6e2410a9e12f', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-17T21:45:34[INFO    ] dripline.core.spimescape(63) -> request received by run_table_service
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(148) -> handling request:{"timestamp": "2016-11-17T21:45:34Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "payload": {"values": [], "run_name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"}}
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(154) -> routing key specifier is: do_insert
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-17T21:45:34Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "payload": {"values": [], "run_name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"}}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-17T21:45:34Z', u'payload': {u'values': [], u'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'6e2410a9e12f', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-17T21:45:34[INFO    ] dripline.core.endpoint(159) -> got a 9 request: {u'values': [], u'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(176) -> method is: <bound method InsertDBEndpoint.on_cmd of <dragonfly.implementations.daq_db_interface.InsertDBEndpoint object at 0x7f9be8be9410>>
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(125) -> operation allowed because not locked
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(179) -> args are:
[]
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(180) -> kwargs are:
{'routing_key_specifier': 'do_insert', u'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(258) -> args are: ()
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(259) -> kwargs are: {'routing_key_specifier': 'do_insert', u'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}
2016-11-17T21:45:34[WARNING ] dragonfly.implementations.daq_db_interface(83) -> unknown error while working with sql
2016-11-17T21:45:34[ERROR   ] dripline.core.endpoint(193) -> got an error: (psycopg2.DataError) value too long for type character varying(80)
 [SQL: 'INSERT INTO runs.run_table (run_name, daq_id) VALUES (%(run_name)s, %(daq_id)s) RETURNING runs.run_table.run_id, runs.run_table.start_timestamp'] [parameters: {'daq_id': 1, 'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}]
2016-11-17T21:45:34[ERROR   ] dripline.core.endpoint(194) -> traceback follows:
Traceback (most recent call last):
  File "/code/dripline/python/dripline/core/endpoint.py", line 181, in handle_request
    result = endpoint_method(*these_args, **these_kwargs)
  File "/code/dripline/python/dripline/core/endpoint.py", line 269, in on_cmd
    result = getattr(self, method_name)(*args, **kwargs)
  File "/code/dragonfly/dragonfly/implementations/daq_db_interface.py", line 135, in do_insert
    self._return_names,
  File "/code/dragonfly/dragonfly/implementations/daq_db_interface.py", line 71, in _insert_with_return
    insert_result = ins.execute()
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/sql/base.py", line 386, in execute
    return e._execute_clauseelement(self, multiparams, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1998, in _execute_clauseelement
    return connection._execute_clauseelement(elem, multiparams, params)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1010, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1146, in _execute_context
    context)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1341, in _handle_dbapi_exception
    exc_info
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1139, in _execute_context
    context)
  File "/usr/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 450, in do_execute
    cursor.execute(statement, parameters)
DataError: (psycopg2.DataError) value too long for type character varying(80)
 [SQL: 'INSERT INTO runs.run_table (run_name, daq_id) VALUES (%(run_name)s, %(daq_id)s) RETURNING runs.run_table.run_id, runs.run_table.start_timestamp'] [parameters: {'daq_id': 1, 'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}]

2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(197) -> request method execution complete
2016-11-17T21:45:34[INFO    ] dripline.core.service(573) -> sending a reply
2016-11-17T21:45:34[INFO    ] dripline.core.service(580) -> reply sent
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(200) -> reply sent
2016-11-17T21:45:34[INFO    ] dripline.core.spimescape(71) -> request processing complete
{}
2016-11-17T21:45:34[INFO    ] dripline.core.service(329) -> Ready for next message
```
3. The respective daq interface service logs:
```
2016-11-17T21:45:33[INFO    ] dripline.core.service(316) -> received a message
2016-11-17T21:45:33[DEBUG   ] dripline.core.service(363) -> Acknowledging message 1
2016-11-17T21:45:33[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-17T21:45:33Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "payload": {"values": ["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", 120]}}
2016-11-17T21:45:33[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-17T21:45:33Z', u'payload': {u'values': [u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab', 120]}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'6e2410a9e12f', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-17T21:45:33[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-17T21:45:33[INFO    ] dripline.core.spimescape(63) -> request received by generic_daq
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(148) -> handling request:{"timestamp": "2016-11-17T21:45:33Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "payload": {"values": ["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", 120]}}
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(154) -> routing key specifier is: start_timed_run
2016-11-17T21:45:33[DEBUG   ] dripline.core.message(166) -> original msg was: {"timestamp": "2016-11-17T21:45:33Z", "msgop": 9, "msgtype": 3, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "payload": {"values": ["aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab", 120]}}
2016-11-17T21:45:33[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'timestamp': u'2016-11-17T21:45:33Z', u'payload': {u'values': [u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab', 120]}, u'msgtype': 3, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'', u'hostname': u'6e2410a9e12f', u'version': u'v0.0.0', u'commit': u''}, u'msgop': 9}
2016-11-17T21:45:33[DEBUG   ] dripline.core.message(148) -> msgtype is 3
2016-11-17T21:45:33[INFO    ] dripline.core.endpoint(159) -> got a 9 request: {u'values': [u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab', 120]}
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(176) -> method is: <bound method DAQProvider.on_cmd of <dragonfly.implementations.daq_run_interface.DAQProvider object at 0x7f0abc633750>>
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(125) -> operation allowed because not locked
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(179) -> args are:
[u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab', 120]
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(180) -> kwargs are:
{'routing_key_specifier': 'start_timed_run'}
2016-11-17T21:45:33[DEBUG   ] dripline.core.endpoint(258) -> args are: (u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab', 120)
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(259) -> kwargs are: {'routing_key_specifier': 'start_timed_run'}
2016-11-17T21:45:34[DEBUG   ] dripline.core.service(525) -> request to send to <run_table.do_insert> is: {
    "timestamp": "2016-11-17T21:45:34Z", 
    "sender_info": {
        "username": "root", 
        "exe": "/usr/local/bin/dragonfly", 
        "package": "dripline", 
        "service_name": "", 
        "hostname": "6e2410a9e12f", 
        "version": "v0.0.0", 
        "commit": ""
    }, 
    "payload": {
        "values": [], 
        "run_name": "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab"
    }, 
    "msgop": 9
}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(166) -> original msg was: {"msgtype": 2, "sender_info": {"username": "root", "exe": "/usr/local/bin/dragonfly", "package": "dripline", "service_name": "run_table_service", "hostname": "6e2410a9e12f", "version": "v0.0.0", "commit": ""}, "timestamp": "2016-11-17T21:45:34Z", "payload": {"values": [null]}, "return_msg": "(psycopg2.DataError) value too long for type character varying(80)\n [SQL: 'INSERT INTO runs.run_table (run_name, daq_id) VALUES (%(run_name)s, %(daq_id)s) RETURNING runs.run_table.run_id, runs.run_table.start_timestamp'] [parameters: {'daq_id': 1, 'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}]", "retcode": 999}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(139) -> from_dict is: {u'retcode': 999, u'timestamp': u'2016-11-17T21:45:34Z', u'return_msg': u"(psycopg2.DataError) value too long for type character varying(80)\n [SQL: 'INSERT INTO runs.run_table (run_name, daq_id) VALUES (%(run_name)s, %(daq_id)s) RETURNING runs.run_table.run_id, runs.run_table.start_timestamp'] [parameters: {'daq_id': 1, 'run_name': u'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaab'}]", u'msgtype': 2, u'payload': {u'values': [None]}, u'sender_info': {u'username': u'root', u'exe': u'/usr/local/bin/dragonfly', u'package': u'dripline', u'service_name': u'run_table_service', u'hostname': u'6e2410a9e12f', u'version': u'v0.0.0', u'commit': u''}}
2016-11-17T21:45:34[DEBUG   ] dripline.core.message(148) -> msgtype is 2
2016-11-17T21:45:34[ERROR   ] dripline.core.endpoint(193) -> got an error: 999
2016-11-17T21:45:34[ERROR   ] dripline.core.endpoint(194) -> traceback follows:
Traceback (most recent call last):
  File "/code/dripline/python/dripline/core/endpoint.py", line 181, in handle_request
    result = endpoint_method(*these_args, **these_kwargs)
  File "/code/dripline/python/dripline/core/endpoint.py", line 269, in on_cmd
    result = getattr(self, method_name)(*args, **kwargs)
  File "/code/dragonfly/dragonfly/implementations/daq_run_interface.py", line 176, in start_timed_run
    self.start_run(run_name)
  File "/code/dragonfly/dragonfly/implementations/daq_run_interface.py", line 106, in start_run
    self.run_name = run_name
  File "/code/dragonfly/dragonfly/implementations/daq_run_interface.py", line 88, in run_name
    result = self.provider.cmd(self.run_table_endpoint, 'do_insert', payload={'run_name':value})
  File "/code/dripline/python/dripline/core/provider.py", line 149, in cmd
    return self._send_request(**request_args)
  File "/code/dripline/python/dripline/core/provider.py", line 116, in _send_request
    raise exception_map[reply.retcode](reply.return_msg, result=reply.payload)
KeyError: 999

2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(197) -> request method execution complete
2016-11-17T21:45:34[INFO    ] dripline.core.service(573) -> sending a reply
2016-11-17T21:45:34[INFO    ] dripline.core.service(580) -> reply sent
2016-11-17T21:45:34[DEBUG   ] dripline.core.endpoint(200) -> reply sent
2016-11-17T21:45:34[INFO    ] dripline.core.spimescape(71) -> request processing complete
{}
2016-11-17T21:45:34[INFO    ] dripline.core.service(329) -> Ready for next message
```
and when attempting to _do_post_run_gets it crashes due to missing attributes in DAQProvider:
```
2016-11-17T21:47:34[INFO    ] dragonfly.implementations.daq_run_interface(127) -> doing postrun snapshot gets
2016-11-17T21:47:34[CRITICAL] dripline.core.service(435) -> Service <generic_daq> crashing with error message:
'DAQProvider' object has no attribute '_start_time'
2016-11-17T21:47:34[ERROR   ] dripline.core.service(436) -> Traceback (most recent call last):
  File "/code/dripline/python/dripline/core/service.py", line 433, in run
    self._connection.ioloop.start()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 138, in start
    self.poller.start()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 434, in start
    self.process_timeouts()
  File "/usr/local/lib/python2.7/site-packages/pika/adapters/select_connection.py", line 283, in process_timeouts
    callback()
  File "/code/dragonfly/dragonfly/implementations/daq_run_interface.py", line 93, in end_run
    self._do_postrun_gets()
  File "/code/dragonfly/dragonfly/implementations/daq_run_interface.py", line 130, in _do_postrun_gets
    snapshot_result = self.provider.cmd(target, 'get_logs', [self._start_time,time_now], timeout=120)
AttributeError: 'DAQProvider' object has no attribute '_start_time'

2016-11-17T21:47:34[DEBUG   ] dripline.core.service(404) -> loop ended
```

Appropriate error handling must be done in both RunDBInterface/InsertDBEndpoint and DAQProvider.

**Comment by les67** [comment]
This issue is now solved in bug/DatabaseIssues branch. Pull request will be submitted when the related issues are solved as well.

**Comment by les67** [comment]
Resolved with pull request #80.
